### PR TITLE
Read tar from chunks on import

### LINKED ISSUE #4545: Read chunked imports on the fly instead of reassembling the file

### PR DESCRIPTION
**Pull request by mdellweg**
fixes #4545

~[noissue]~

### CODE REVIEW

**Comment by ggainey** [comment]
If we can avoid reassembling, it saves complexity in a number of places. Example: responding to a recent bug, I had to teach the import-code to recognize that, while the TOC says "chunked", the recombined file is all that we have left, so let's check *its* csum instead of failing due to "chunks missing". 

Let's pursue this and see where it gets us.

**Comment by dralley** [comment]
I'm not against this, especially if Grant likes the idea, but the reconstruction just isn't *that* expensive in the grand scheme of the import, which (it sounds like) we've already gotten into a reasonable state before most of the optimizations we're working on are even merged.  And it looks like a lot of complexity in itself.

**Comment by dralley** [comment]
@mdellweg Could you fix up the merge conflict?

I'll try this out tomorrow

**Comment by mdellweg** [comment]
> I'm not against this, especially if Grant likes the idea, but the reconstruction just isn't _that_ expensive in the grand scheme of the import, which (it sounds like) we've already gotten into a reasonable state before most of the optimizations we're working on are even merged. And it looks like a lot of complexity in itself.

It's not so much about speed here, though i guess we can gain something, because we do not need to write a huge file to a spinny disk (i know last century technology...). My main motivation is to remove a step in the import process (and a lot of complexity with it), that deletes input files and endangers the import process that way. The additional bonus: We can import from a read-only volume that way.

**Comment by dralley** [comment]
Well, it does work and it does knock a sizable chunk off the front-end of the import process (about half, although I imagine it would vary a bit - note that's just the first task and not the whole import)

After taking a closer look at the code I'm more OK with it than I was originally.  I'll review more properly tomorrow.

**Review by lubosmj — APPROVED** [review]
As per my testing, this does not affect standard pulp-container workflows with chunked exports. I am approving.

**Review by ggainey — CHANGES_REQUESTED** [review]

**Comment by ianballou** [comment]
From the Katello side, can we have some general info on how the files on disk will differ with this change? As per @parthaa we don't do cleanup for users and cleanup isn't documented, so we might not have anything to change if the file structure ends up being different.

**Comment by dralley** [comment]
@ianballou In the event that the user is using the chunked option, the chunks will not be consolidated during the import process.  

You start with N chunks of some amount of data on some media and a tble of contents JSON file, and you either copy it to a location on the host or transfer the data from the media to some location on the host.  Then you import from that directory containing the chunks and the table of contents JSON file. 

* currently, those N chunks will be deleted and replaced with one large file reconstructed from the N smaller chunks
* with this PR, there will be no large file created and no chunks deleted, you will still have exactly the same data present which was copied in the first place

of having one file with hundreds of gigabytes / >1Tb of information, you will retain the same N chunks you moved over to the system in the first place.

File: pulpcore/app/tasks/importer.py [hunk 1]
```diff
@@ -57,6 +57,132 @@
...
+        ]
+        self.chunk = 0
+        self.offset = 0
+        return self
+
+    def __exit__(self, *exc):
+        super().__exit__(*exc)
+        del self.chunks
+        del self.chunk
+        del self.offset
```
[the remark this answers is not in the document]
Comment (mdellweg) [reply pulpcore/app/tasks/importer.py:100]: Probably not necessary, i thought it'd be a safety measure. Any one of `tell`, `read` and `seek` will fail that way when used outside of the context manager.

**Comment by dralley** [comment]
So to be more specific about the performance:

If I export the RHEL8 baseos repo and then import it:

Previously, the `pulp_import` task would take ~800 seconds and the `import_repository_version` task would take ~1900 seconds

With this PR, the `pulp_import` task takes about ~300 seconds and the `import_repository_version` task remains the same.

Including my [new PR](https://github.com/pulp/pulpcore/pull/4588) on top of this,  the `pulp_import` task takes about ~300 seconds and the `import_repository_version` task takes about ~550 seconds.

So overall that's a 2/3rds drop.

**Review by dralley — COMMENTED** [review]
LGTM overall.  However, I would really like to have some unit tests for basic reading and seeking if possible.  Maybe validating too, but I don't care as much about that.

**Comment by dralley** [comment]
I'll file an issue for additional unit tests.

**Pull request merged**